adds link to ZK Stack (#634)
matter-labs/zksync-web-era-docs: docs/reference/concepts/hyperscaling.md

7. The receiving Hyperchain settles its proof, where the imported transaction root is also verified.  

<div  align="center">

![](../../assets/images/hyperscalingBridgingFull.png)

</div>

There are three types of bridges in this ecosystem. The enshrined L1-L2 bridge, the [zkPorter](https://blog.matter-labs.io/zkporter-a-breakthrough-in-l2-scaling-ed5e48842fbf) shard bridges, and the hyperbridges will all operate with a similar set of interfaces. 

Hyperbridges technically will be similar to L2→ L1 bridges, they will be asynchronous and not atomic. However, with the help of Account Abstraction, external relayers, and lower fees on rollups, the users will not have to initiate the call on the destination chain. This means the user experience will feel like an L1 → L2 bridge. You can try learn more about L1<>L2 bridging [here](https://era.zksync.io/docs/dev/how-to/send-message-l2-l1.html).

Besides cross-chain calls, it will be also possible to execute cross-chain views inside transactions, accessing arbitrary data from other chains in the ecosystem. 

### Hyperchain User Experience

> As we aim to onboard billions of users to Web3, user experience has to be a priority. Users will have wallets on different chains, unified by cross-chain wallet management. This is still a hot research topic, as seen in Vitalik's [recent post](https://vitalik.ca/general/2023/06/20/deeperdive.html). These solutions will be integrated into [Account Abstractions](https://era.zksync.io/docs/reference/concepts/aa.html), which are natively supported on our chains.

Author: ```suggestion
As we aim to onboard billions of users to Web3, user experience has to be a priority. Users will have wallets on different chains, unified by cross-chain wallet management. This is still a hot research topic, as seen in Vitalik's [recent post](https://vitalik.ca/general/2023/06/20/deeperdive.html). These solutions will be integrated into [Account Abstractions](https://era.zksync.io/docs/reference/concepts/account-abstraction.html), which are natively supported on our chains.
```

Here proof aggreagation happens via the L2, as the proofs of different L2 blocks are aggregated when settling on L1. This method is ultimately not scalable, as the L2's VM will be a bottleneck for proof verification. The L2's VM will also require a full consensus mechanism, meaning long term storage, transaction verification etc.

<div  align="center">

![](../../assets/images/hyperscalingL3Fast2Blocks.png)

</div>

### Layered Aggregation 

Layered Aggregation combines the benefits of L3s with the benefits of aggregation. The L2's VM is replaced by a specialised proof that still permits messaging and aggregation. This proof tracks the State Root of the participating rollups, as well as the Transaction Root. The Transaction Root will be imported from and settled inside this specialised proof. Compared to the L2's VM this solution is more scalable, and will only need a lightweight consensus mechanism. 

<div  align="center">

![](../../assets/images/hyperscalingLayeredAggregation.png)

</div>

### Economic Guarantees

> The time it takes to hyperbridge is the settlement time of the proof, which can only happen after the proof is generated. This will be on the order of minutes, but for some chains faster interoperability will be needed. This is impossible to achieve trustlessly, but it is easy to achieve if have economic guarantees. The transaction root can be calculated outside the proof and then imported. This is not as secure as importing the transaction root from a proof, even a single invalid transaction means that all of the participating rollups will have to revert, as valid proofs cannot be generated. So we are building this solution as an optional add-on, and not part of the core protocol. 

Author: Typo here? "but it is easy to achieve if have economic guarantees."

<div  align="center">

![](../../assets/images/hyperscalingFastEconomic.png)

</div>

### Atomicity

Another property that we would like to enable long-term is atomicity for the transactions. This can be achieved using the DA layer. Instead of a relayer sending the transaction to the destination rollup, it can be posted to the DA layer, and the receiving rollup can read it from there. To guarantee atomicity, the received transactions need to be compared to the sent transactions. If rollups generated blocks at the same frequency then this could be done from block to block inside the proof aggregator. Unfortunately this is not the case, rollups will choose to generate blocks and setle proofs at their own rate. This makes the mechanism more complicated, a detailed description is available [here](https://hackmd.io/@kalmanlajko/BkmTTADai), and is still under review. This will only be possible for the Layered Aggregator. 

<div  align="center">

![ ](../../assets/images/hyperscalingAtomicityDA.png)

> This image is an intuition pump, the real mechanism is more complicated and works differently. 

Author: shall we remove this phrase or amend it to something like "is a high level overview"

</div>

### Sovereignty

All Hyperchains will be sovereign in the ecosystem. This means two things.

> Proof aggregation will be optional, Hyperchains can choose to not participate. In this case, they can settle their proofs directly to Ethereum for a much larger fee. Aggregation will also be decentralized and widely accessible, meaning the hardware requirements to run a prover will be as low as possible.

Author: ```suggestion
1. Proof aggregation will be optional, Hyperchains can choose to not participate. In this case, they can settle their proofs directly to Ethereum for a much larger fee. Aggregation will also be decentralized and widely accessible, meaning the hardware requirements to run a prover will be as low as possible.
```

<div  align="center">

![](../../assets/images/hyperscalingSovereignty.png)

</div>

> Second, Hyperchains will be able to permissionlessly join and exit the ecosystem, adding or removing all their assets to the common pool in the Shared Bridge. Joining is self-explanatory, everyone will have the right to boot up new Hyperchains and join the ecosystem in a Chain Factory contract.

Author: ```suggestion
2. In addition, Hyperchains will be able to permissionlessly join and exit the ecosystem, adding or removing all their assets to the common pool in the Shared Bridge. Joining is self-explanatory, everyone will have the right to boot up new Hyperchains and join the ecosystem in a Chain Factory contract.
```

Exiting will usually not be a similarly wise decision, as interoperability will be lost with other Hyperchains. However, the ecosystem could sometimes upgrade due to governance, and in this case, it is imperative that each Hyperchain have the right to rage quit. In this case, there will be a mandatory upgrade period during which the Hyperchains that disagrees with the upgrade can exit in a coordinated fashion.

### Feature comparison 

We look at the comparisons that different aggregator mechanisms enable. 

<div  align="center">

|   | Aggregation | L3s | Layered Aggregation  |
|---|---|---|---|
|Fast Messaging|  No | Yes  |  Yes |
| Scales   | Yes  |  No | Yes  |
| Consensus Mechanism | None  | L2 Full Consensus  | Lightweight Consensus  |
| Instant Messaging Add-on  | No  | Yes | Yes |
| Hyperbridges atomic | No | No | Yes via DA, under review|
| Sovereign | Yes  | Yes  | Yes |

</div>

> ## Modularity: Hyperchain Customization

Author: ```suggestion
## ZK Stack: Hyperchain Customization
```
just to mention ZK Stack more

# Hyperscaling

## Scaling Blockchains

Today, Ethereum processes around 12 transactions per second. To support a future in which the world's financial activity is on-chain, we need to be able to process millions of transactions per second. However, blockchains are hard to scale due to their decentralized nature. The holy grail has always been horizontal scalability. Similar to the internet, where different servers host different websites, multiple distinct chains could host different applications in parallel. There have been many experiments attempting horizontal scaling, such as Polkadot, Cosmos, Near, and even the now-outdated Eth 2.0 vision, all of which have multiple shards or multiple chains. All of these solutions take care of interoperability, as the shards and bridges are designed together, as both are integral to the ecosystem. Unfortunately, the shards could never fully trust each other in any of these solutions.

<div  align="center">

![ ](../../assets/images/hyperscalingCosmos.png)

The Cosmos ecosystem. Unfortunately, the bridges have trust requirements, the ecosystem ultimately does not scale. 

</div>

> In parallel, the Ethereum community was solving the trust problem by verifying off-chain computations on-chain, inventing Plasma, optimistic rollups, and other solutions. It is now clear that zero-knowledge proofs are the future, as they provide cryptographic security. Ethereum will be accessible to everybody. The [DA layer](https://ethereum.org/en/roadmap/danksharding/) and ZK Rollups together will provide a secure and scalable ecosystem. But how will the rollups interoperate? Now that we have figured out the building blocks, it is time to review the system as a coherent whole again.

Reviewer: I suggest shortening this introduction to something as follows and remove the cosmos image:

Ethereum, currently processing around 12 transactions per second, must scale to handle millions of transactions for a future of global on-chain financial activities. Achieving this in a decentralized structure poses significant challenges. Various solutions, including Polkadot, Cosmos, Near, and Eth 2.0, have been explored with multi-chain or shard architectures, but full trust remains an issue. The widely accepted solution for these challenges lies in zero-knowledge proofs, offering cryptographic security. Combined with the DA layer and ZK Rollups, these can create a secure, scalable ecosystem, making Ethereum accessible to all.

## The Problem: Current bridging solutions

The current state of bridging solutions is a [mess](https://twitter.com/VitalikButerin/status/1479501366192132099). There are some good solutions, such as atomic swaps, which enable trustless asset swapping between different blockchains. Unfortunately, they are not suitable for general message passing.

<div  align="center">

> ![](../../assets/images/hyperscalingIslands.png)

Reviewer: I think this image can be removed.

```suggestion
```

</div>

In order to pass information between chains, bridges based on economic incentives were introduced. These require the operators to lock funds, which can be slashed based on an arbitration process.

The potential design space is vast, and there are several potential issues. These include:

1. The arbitration process is not trustless; dapp developers have to consider the risks involved.
2. These bridges usually do not transfer real assets cross-chain; they only free up already locked funds. This makes the process expensive. However, this can be solved by using general message passing and minting synthetic assets. But the synthetic assets are less secure than the original assets and are not interchangeable with them.

> 3. Chains can hard fork. This is their right, but it endangers the entire bridging ecosystem. This can be mitigated by locking funds separately for each bridge, but this makes the process much more expensive due to liquidity fragmentatio This applies to ZK-enabled bridges as well.

Reviewer 2: ```suggestion
3. Chains can hard fork. This is their right, but it endangers the entire bridging ecosystem. This can be mitigated by locking funds separately for each bridge, but this makes the process much more expensive due to liquidity fragmentation. This applies to ZK-enabled bridges as well.
```

4. Finally, even if the economic incentives are sound, hacks can always occur. If a vulnerability is found in the smart contracts, then every bridge in the entire ecosystem is affected. With the growth of adoption, cross-chain activity will be an indispensable part of every chain's activity. At this stage, if a bug is found, then every chain will want to hard fork to save its bridges. Bridges can only be hard forked on both sides with consensus. This means there needs to be a common social consensus for the ecosystem.

## The Solution: Hyperchains + Hyperbridges

> 

Author: I think should mention what Hyperchains are in this section. We can bring the original section ["What are hyperchains"](https://era.zksync.io/docs/reference/concepts/hyperscaling.html#what-are-hyperchains) here. It will also help with SEO 😉

### Sovereignty

All Hyperchains will be sovereign in the ecosystem. This means two things.

Proof aggregation will be optional, Hyperchains can choose to not participate. In this case, they can settle their proofs directly to Ethereum for a much larger fee. Aggregation will also be decentralized and widely accessible, meaning the hardware requirements to run a prover will be as low as possible.

<div  align="center">

![](../../assets/images/hyperscalingSovereignty.png)

</div>

Second, Hyperchains will be able to permissionlessly join and exit the ecosystem, adding or removing all their assets to the common pool in the Shared Bridge. Joining is self-explanatory, everyone will have the right to boot up new Hyperchains and join the ecosystem in a Chain Factory contract.

Exiting will usually not be a similarly wise decision, as interoperability will be lost with other Hyperchains. However, the ecosystem could sometimes upgrade due to governance, and in this case, it is imperative that each Hyperchain have the right to rage quit. In this case, there will be a mandatory upgrade period during which the Hyperchains that disagrees with the upgrade can exit in a coordinated fashion.

> ### Conclusion 

Author: ```suggestion
### Feature comparison 
```
Just cause "Conclusion" in the middle of the page does not feel right

We look at the comparisons that different aggregator mechanisms enable. 

<div  align="center">

|   | Aggregation | L3s | Layered Aggregation  |
|---|---|---|---|
|Fast Messaging|  No | Yes  |  Yes |
| Scales   | Yes  |  No | Yes  |
| Consensus Mechanism | None  | L2 Full Consensus  | Lightweight Consensus  |
| Instant Messaging Add-on  | No  | Yes | Yes |
| Hyperbridges atomic | No | No | Yes via DA, under review|
| Sovereign | Yes  | Yes  | Yes |

</div>

> # Modularity: Hyperchain Customization

Author: ```suggestion
## Modularity: Hyperchain Customization
```
To create an H2 that will show in the table of contents on the side

> The main customization options to be provided by [ZK Stack (Post)](https://www.notion.so/ZK-Stack-Post-9d79299086cc4f11acdb78445b8b523f?pvs=21) is explained below. Developers are of course free to implement their own components and customizations.

Author: ```suggestion
The main customization options to be provided by [ZK Stack (Post)](https://blog.matter-labs.io/introducing-the-zk-stack-c24240c2532a) is explained below. Developers are of course free to implement their own components and customizations.
```
probably this link instead

> ## **Sequencing transactions**

Author: ```suggestion
### Sequencing transactions
```

- **Decentralized sequencer** - In this mode, a Hyperchain will coordinate on what transactions are included in a block using a consensus algorithm. It can be any algorithm, so developers can reuse existing implementations (e.g. Tendermint or HotStuff with permissionless dPoS). But we can also take advantage of the fact that finality checkpoints are guaranteed by the underlying L1, and implement an algorithm that is simpler and boasts higher performance. The Basechain will switch to this option as soon as the consensus implementation is ready, and will make its code available to the Hyperchain developers.
- **Priority queue** - This simply means absence of any sequencer: all transactions can be submitted in batches via the priority queue from an underlying L2 or even L1 chain, taking advantage of their stronger censorship-resistance. It might be especially interesting for special-purpose governance protocols (e.g. on-chain voting). It’s worth noting that the priority queue will always be available as an escape-hatch mechanism (even if a centralized or decentralized sequencer is employed), to protect users against censorship by a malicious sequencer.

> ## **Data availability**

Author: ```suggestion
### Data availability
```

> ## **Logical state partitions in ZK Porters**

Author: ```suggestion
### Logical state partitions in ZK Porters
```

Each Hyperchain can have one or more logical partitions that are part of the same state but live in separate subtrees and enforce different data availability policies. From the user perspective, they would appear as separate Hyperchain instances (with their own chain ID, separate wallet connection, block explorer view, etc), which can however interoperate synchronously.

Synchronicity is important as it enables atomic transactions between partitions, unlocking several unique use cases:

- Transparently reading the state of another partition.
- Using flash loans between the partitions.

One prominent example of this is a combination of **[zkRollup + zkPorter](https://blog.matter-labs.io/zkporter-a-breakthrough-in-l2-scaling-ed5e48842fbf)** (which will be part of the zkSync Basechain):

![](../../assets/images/hyperscalingZKPorter.png)

> ## **Privacy**

Author: ```suggestion
### Privacy
```
